trustfall_core/src/interpreter/execution.rs

```diff
@@ -286,10 +288,52 @@ fn get_max_fold_count_limit(carrier: &mut QueryCarrier, fold: &IRFold) -> Option
     result
 }
 
+/// If this IRFold has a filter on the folded element count, and that filter imposes
+/// a min size that can be statically determined, return that min size so it can
+/// be used for further optimizations. Otherwise, return None.
+fn get_min_fold_count_limit(carrier: &mut QueryCarrier, fold: &IRFold) -> Option<usize> {
+    let mut result: Option<usize> = None;
+
+    let query_arguments = &carrier.query.as_ref().expect("query was not returned").arguments;
+    for post_fold_filter in fold.post_filters.iter() {
+        let next_limit = match post_fold_filter {
+            // We do not need to visit Equals and OneOf here,
+            // since those will be handled by `get_max_fold_count_limit`
+            Operation::GreaterThanOrEqual(
+                FoldSpecificFieldKind::Count,
+                Argument::Variable(var_ref),
+            ) => {
+                let variable_value = query_arguments[&var_ref.variable_name].as_usize().unwrap();
+                Some(variable_value)
+            }
+            Operation::GreaterThan(FoldSpecificFieldKind::Count, Argument::Variable(var_ref)) => {
+                let variable_value = query_arguments[&var_ref.variable_name].as_usize().unwrap();
+                Some(variable_value + 1)
+            }
+            _ => None,
+        };
+
+        match (result, next_limit) {
+            (None, _) => result = next_limit,
+            (Some(l), Some(r)) if l < r => result = next_limit,
+            _ => {}
+        }
+    }
+
+    result
+}
+
 fn collect_fold_elements<'query, Vertex: Clone + Debug + 'query>(
     mut iterator: ContextIterator<'query, Vertex>,
     max_fold_count_limit: &Option<usize>,
+    min_fold_count_limit: &Option<usize>,
```

Comment:
if we make this an isize instead of a usize, then we can make -1 the None case and save half the space, going from 16 for the Option to 8 for an isize, but it might be slightly more confusing.

Reply:
I think that's not worth it, too much likelihood of introducing weird bugs.